### ISSUE TITLE
USB passthrough to lxc

### ISSUE THREAD

**Issue by nubis** [maruos/maruos#29]
Hi guys,

I'm still learning about the architecture used here, as I see you're using lxc containers. Is there any way to just pass anything in the USB bus to the container?

I'm trying to use an OTG device and I see them listed in the kernel but not usable by the container, (lsusb bails out with a -99 error).

I think I should be able to play around with the lxc config without building my own maruOS image and reinstalling. Is that correct? If that's the case, where can I find that config? Also, I probably need to root my phone to play with those files live without building my own image, right?

**Comment by pdsouza** [comment]
Yep, you can make edits to the LXC configuration file without building your own image--all you need is root access to edit the file. Take a look at `/data/maru/containers/jessie/config`. 

At the moment, only the bare minimum set of devices (`/dev/input` for BT keyboard and mouse) are mounted into the container.

You can try mounting extra devices you need but I haven't tried USB myself yet. You may need to play around with which device node paths you mount since they may not exist before the USB device is plugged in.

**Comment by nubis** [comment]
I've got my USB devices working via an OTG cable, so I wanted to post what I've learned as a micro tutorial in case anyone else needs it.
I strongly recommend just using wireless network storage and bluetooth for peripherals, but in my case I have a Bitcoin Trezor that I need to use, and that's USB only (it registers as a HID Human Interaction Device over an USB char buffer).
Also, you'll need to use VNC to see your screen after you unplug your slimport to plug in the
OTG cable. You don't get use slimport and OTG simultaneously, at least not on the nexus 5.

So, without further ado, here's what you can do to get USB working:

1) Root your phone. You'll need to do this to allow the LXC container which runs MaruOS to use the
USB devices. I just used the TWRP root option.

2) Install the "Terminal Emulator" app, (by Jack Palevich) on Android.

3) Install the "VNC Viewer" app, (by RealVNC Limited) on Android.

4) Install usbutils and x11vnc on debian and run it. It prompts you to set a password the first time. (man x11vnc to learn more)
```
sudo apt-get install usbutils x11vnc
x11vnc -usepw -localhost -display :0 -forever
```

5) Now the fun part, you need to find your device on /dev on the android side.
a) Unplug your slimport cable and plug in your OTG cable and USB device.
b) Launch the "Terminal Emulator" app on Android and explore your /dev looking
for your device nodes. My session was something like this:
```
$ su 
# cd /dev/bus/usb/001
# ls -al
crw------- root     root     189,     0 1970-01-01 21:00 001 
crw------- root     root     189,     1 1970-01-01 21:00 002 
```

c) Take notice of the following:
  - The full paths were /dev/bus/usb/001/001 and /dev/bus/usb/001/002.
  - The initial 'c' in 'crw------' means this is a 'char' device, a 'b' would mean a 'block' device.
  - The "189, 0" and "189, 1", those are the "addresses" of your device, the OTG hub, etc.

d) Keep digging and repeat the previous step. Look for 'b' nodes in /dev/block/sda1 if your device
is a thumb drive or other type of storage.

6) Allow the Maru LXC container to access the devices you discovered.
You do this by editing the LXC configuration file: /data/maru/containers/jessie/config
After digging around in my device I decided I wanted to let LXC access anything in the
189 and 180 ranges which were what my Trezor changed.
Since you don't have a text editor on android you can edit the file like so:
```
su
cat "lxc.cgroup.devices.allow = c 189:* rwm" >> /data/maru/containers/jessie/config
cat "lxc.cgroup.devices.allow = c 180:* rwm" >> /data/maru/containers/jessie/config
```
rwm stands for Read, Write, Mount. (I think). The 'c' is for what we looked up before, a char
device.

7) Now, plug your slimport back in and make sure you replicate the directory structure you saw in /dev on the Android side inside the LXC container. So, on the debian side, using mkdir and mknod, I ended up doing something
like this:
```
sudo mkdir --parents /dev/bus/usb/001
sudo mkdir --parents /dev/bus/usb/002
sudo mknod /dev/bus/usb/001/001 c 189 0
sudo mknod /dev/bus/usb/001/002 c 189 1
sudo mknod /dev/bus/usb/002/001 c 189 128
sudo mkdir --parents /dev/usb
sudo mknod /dev/usb/hiddev0 c 180 96
sudo ln -s /dev/usb/usb1-1:1.0 /dev/usb/hiddev0
sudo mkdir --parents /dev/block
sudo mknod /dev/block/sda b 8 0
sudo mknod /dev/block/sda1 b 8 1
sudo mknod /dev/block/sda2 b 0 2
```
Your mileage may vary so be careful to build this structure the same as you have it on the
Android side. I think I've overdone it though. Also, the mknod won't survive a reboot, you may want
to make this into a shell script to be run whenever you need to use your OTG stuff, or run them at debian boot. It still works if you do the 'mknod' before plugging in the device.

8) Plug in your USB device again. Use VNC viewer to see your debian desktop on your phone screen.
Try running 'lsusb' as root on a debian terminal and see if your device is listed, if so, you're done.

----

Well that's it, there's a lot of cleaning up to do, ideally you'll set your permissions such that
you don't need to use USB as root.

**Comment by pdsouza** [comment]
@nubis Awesome work! Thank you so much for taking the time to write this great tutorial.

I think we can add those "lxc.cgroup.devices.allow" lines for device numbers 180 and 189 into Maru's default container config: https://github.com/maruos/vendor_maruos/blob/master/container/jessie/config. That'll be one less step to get USB access working on Maru!

**Comment by nubis** [comment]
Yeah, that would cut down a bunch of steps. You would just need the mknod parts.

I've learned just enough about USB and udev to get this working. I think device numbers are bound to change so whitelisting 180 and 189 may be brittle or may need to be changed when porting to other phones.

Is it too insecure to just use: "lxc.cgroup.devices.allow *:*" ? I mean, I'd personally like debian to be a first class citizen not just a sandbox inside android. Also, I don't know about udev rules but I think those can be set so that the nodes are created automatically whenever a device is plugged in.
I'll keep playing around with this and see if I can wrap up a solid pull request.

Given that you still need to see your desktop through that tiny phone screen, I don't think people are going to be using usb stuff unless it's absolutely necessary.

Thanks for your help, keep up the great work :)

**Comment by k3a** [comment]
@nubis rwm stands for read, write and mknod (https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt)
